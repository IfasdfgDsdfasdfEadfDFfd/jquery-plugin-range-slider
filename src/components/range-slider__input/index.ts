export * from './range-slider__input.view';
