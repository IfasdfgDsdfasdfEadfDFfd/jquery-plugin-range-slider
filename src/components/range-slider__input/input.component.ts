import { Component } from '@core';

import { InputController } from './input.controller';
import { InputModel } from './input.model';
import { Input } from './input.view';

class InputComponent extends Component {
  view = new Input();
  model = new InputModel();
  controller = new InputController();
}

export { InputComponent };
