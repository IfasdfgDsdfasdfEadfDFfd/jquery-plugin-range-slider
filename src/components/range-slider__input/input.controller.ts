import { Controller } from '@core';
import { INPUT_ACTIONS } from './input.model';

class InputController extends Controller {
  mapState({ input }: RangeSliderModelData): Partial<InputProps> {
    return input;
  }

  mapDispatch({ inputDispatch: dispatch }: Record<string, ModelDispatch>): Partial<InputProps> {
    return {
      valueChangeHandler(index, value) {
        dispatch({
          type: INPUT_ACTIONS.CHANGE_VALUE,
          payload: { index, value },
        });
      },
    };
  }
}

export { InputController };
