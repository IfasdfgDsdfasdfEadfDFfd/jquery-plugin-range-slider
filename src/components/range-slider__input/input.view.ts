import { View, ContainerView } from '@core';
import { memo } from 'core/utils';

class Input extends ContainerView<InputProps> {
  attrs = {
    class: 'range-slider__input',
  };

  childViewClass = InputElement;

  render(props: InputProps): void {
    super.render(props);

    this.nativeElement.classList.toggle(`${this.attrs.class}_interval`, props.values.length > 1);
  }

  getProps({ values, ...restProps }: InputProps): ContainerViewProps {
    return { iterator: values.map((value, index) => ({ value, index })), restProps };
  }
}

@memo({ methods: ['updateMin', 'updateMax', 'updateStep', 'updateValue', 'updateInputHandler'] })
class InputElement extends View<InputElementProps> {
  tag = 'input';
  attrs = {
    type: 'range',
    class: 'range-slider__input__element',
  };

  render({ min, max, step, value, index, valueChangeHandler }: InputElementProps): void {
    this.updateMin(min);
    this.updateMax(max);
    this.updateStep(step);
    this.updateValue(value);
    this.updateInputHandler(value => valueChangeHandler(index, value));
  }

  updateMin(value: number): void {
    this.nativeElement.setAttribute('min', value.toString());
  }

  updateMax(value: number): void {
    this.nativeElement.setAttribute('max', value.toString());
  }

  updateStep(value: number): void {
    this.nativeElement.setAttribute('step', value.toString());
  }

  updateValue(value: number): void {
    this.nativeElement.setAttribute('value', value.toString());
    (this.nativeElement as HTMLInputElement).value = value.toString();
  }

  updateInputHandler(handler: (value: number) => void): void {
    this.nativeElement.oninput = event => {
      const { value } = event.target as HTMLInputElement;
      handler(parseFloat(value));
    };
  }
}

export { Input, InputElement };
