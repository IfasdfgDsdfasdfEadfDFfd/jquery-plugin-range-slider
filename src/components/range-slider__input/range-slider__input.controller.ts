import { Controller } from '@core';

class InputController extends Controller {
  mapState({ input }: RangeSliderModelData): Partial<InputProps> {
    return input;
  }
}

export { InputController };
