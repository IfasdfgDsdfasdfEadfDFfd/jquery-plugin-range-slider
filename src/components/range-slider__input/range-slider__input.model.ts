import { Model } from '@core';

class InputModel extends Model {
  name = 'input';

  data = {
    min: 0,
    max: 10,
    values: [5, 7],
  };
}

export { InputModel };
