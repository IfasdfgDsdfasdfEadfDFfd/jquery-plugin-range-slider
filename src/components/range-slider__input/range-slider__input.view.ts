import { View } from '@core';
import { memo } from 'core/utils';

class Input extends View<InputProps> {
  attrs = {
    class: 'range-slider__input',
  };

  render({ min, max, values, valueChangeHandler }: InputProps): void {
    values.forEach((value, index) => {
      const childProps = { min, max, value, valueChangeHandler };

      if (!this.children[index]) {
        this.children[index] = new InputElement();
        this.children[index].init(this.nativeElement);
      }

      this.children[index].render(childProps);
    });
  }
}

@memo(['updateMin', 'updateMax', 'updateValue'])
class InputElement extends View<InputElementProps> {
  tag = 'input';
  attrs = {
    type: 'range',
    class: 'range-slider__input__element',
  };

  render({ min, max, value }: InputElementProps): void {
    this.updateMin(min);
    this.updateMax(max);
    this.updateValue(value);
  }

  updateMin(value: number): void {
    this.nativeElement.setAttribute('min', value.toString());
  }

  updateMax(value: number): void {
    this.nativeElement.setAttribute('max', value.toString());
  }

  updateValue(value: number): void {
    this.nativeElement.setAttribute('value', value.toString());
  }
}

export { Input, InputElement };
