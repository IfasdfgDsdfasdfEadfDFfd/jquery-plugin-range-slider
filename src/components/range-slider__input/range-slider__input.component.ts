import { Component } from '@core';

import { InputController } from './range-slider__input.controller';
import { InputModel } from './range-slider__input.model';
import { Input } from './range-slider__input.view';

class InputComponent extends Component {
  view = new Input();
  model = new InputModel();
  controller = new InputController();
}

export { InputComponent };
