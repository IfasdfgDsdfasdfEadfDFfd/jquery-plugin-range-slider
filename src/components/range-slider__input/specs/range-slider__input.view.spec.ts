import { Input, InputElement } from 'components/range-slider__input';

describe('Input', () => {
  let input: Input;

  const props = {
    min: 0,
    max: 10,
    values: [1, 2, 3, 4],
    valueChangeHandler: jest.fn(),
  };

  beforeEach(() => {
    input = new Input();
    input.init();
    input.render(props);
  });

  test('to match snapshot', () => {
    expect(input.nativeElement).toMatchSnapshot();
  });
});

describe('InputElement', () => {
  let inputElement: InputElement;

  const props = {
    min: 0,
    max: 10,
    value: 5,
    valueChangeHandler: jest.fn(),
  };

  beforeEach(() => {
    inputElement = new InputElement();
    inputElement.init();
    inputElement.render(props);
  });

  test('updateMin()', () => {
    const nextMin = 3;
    inputElement.updateMin(nextMin);
    expect(inputElement.nativeElement.getAttribute('min')).toEqual(nextMin.toString());
  });

  test('updateMax()', () => {
    const nextMax = 13;
    inputElement.updateMax(nextMax);
    expect(inputElement.nativeElement.getAttribute('max')).toEqual(nextMax.toString());
  });

  test('updateValue()', () => {
    const nextValue = 7;
    inputElement.updateValue(nextValue);
    expect(inputElement.nativeElement.getAttribute('value')).toEqual(nextValue.toString());
  });

  test('to match snapshot', () => {
    expect(inputElement.nativeElement).toMatchSnapshot();
  });
});
