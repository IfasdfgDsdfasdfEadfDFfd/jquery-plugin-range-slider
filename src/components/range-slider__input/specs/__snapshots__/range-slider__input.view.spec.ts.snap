// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input to match snapshot 1`] = `
<div
  class="range-slider__input"
>
  <input
    class="range-slider__input__element"
    max="10"
    min="0"
    type="range"
    value="1"
  />
  <input
    class="range-slider__input__element"
    max="10"
    min="0"
    type="range"
    value="2"
  />
  <input
    class="range-slider__input__element"
    max="10"
    min="0"
    type="range"
    value="3"
  />
  <input
    class="range-slider__input__element"
    max="10"
    min="0"
    type="range"
    value="4"
  />
</div>
`;

exports[`InputElement to match snapshot 1`] = `
<input
  class="range-slider__input__element"
  max="10"
  min="0"
  type="range"
  value="5"
/>
`;
